src/backend/distributed/executor/tuple_destination.c

```diff
@@ -203,8 +213,20 @@ TupleDestDestReceiverReceive(TupleTableSlot *slot,
 	HeapTuple heapTuple = ExecFetchSlotTuple(slot);
 #endif
 
+	DistributedExecutionStats *executionStats = tupleDestReceiver->executionStats;
+	if (SubPlanLevel > 0 && executionStats != NULL)
+	{
+		uint64 heapTupleSizeByte = HeapTupleHeaderGetDatumLength(heapTuple);
+		executionStats->totalIntermediateResultSize += heapTupleSizeByte;
+		if (CheckIfSizeLimitIsExceeded(executionStats))
```

Comment:
We can possibly have `EnsureIntermediateSizeLimitNotExceeded` that both does the check and error if the check is not satisfied. Something like:

```c
static void EnsureIntermediateSizeLimitNotExceeded(DistributedExecutionStats *executionStats) {
    if (!executionStats) {
         return;
    }
	if (!SubPlanLevel || MaxIntermediateResult < 0)
	{
		return;
	}

	uint64 maxIntermediateResultInBytes = MaxIntermediateResult * 1024L;
	if (executionStats->totalIntermediateResultSize < maxIntermediateResultInBytes)
	{
		return;
	}

		ereport(ERROR, (errmsg("the intermediate result size exceeds "
						   "citus.max_intermediate_result_size (currently %d kB)",
						   MaxIntermediateResult),
					errdetail("Citus restricts the size of intermediate "
							  "results of complex subqueries and CTEs to "
							  "avoid accidentally pulling large result sets "
							  "into once place."),
					errhint("To run the current query, set "
							"citus.max_intermediate_result_size to a higher"
							" value or -1 to disable.")));
}
```
That would possibly simplify the code.